### DISCUSSION TITLE
Derive Diesel traits on enums with multi-column variants

### DISCUSSION THREAD

**Discussion by dspicher** [diesel-rs/diesel#3867] [Q&A] [answered]
I have seen how custom enum types that map onto a single column [can be supported](https://github.com/diesel-rs/diesel/blob/bc263afa1765f650d9c4bf36c40536ff95b9cade/diesel_tests/tests/custom_types.rs#L22-L27) inside a struct that derives `Queryable` and `Insertable`.

In my use-case, however, the enum variants themselves have multiple fields that map onto multiple columns each, something like this (sorry for the contrived example):

```rust
#[derive(diesel::Insertable, diesel::Queryable)]
struct Menu {
    food: String,
    #[diesel(embed)]
    beverage: TeaOrCoffee,
}

enum TeaOrCoffee {
    Tea(Tea),
    Coffee(Coffee),
}

struct Tea {
    temperature: u8,
    kind: String,
}

struct Coffee {
    roast: String,
    origin: String,
}

diesel::table! {
    menus {
        id -> Integer,
        food -> Text,
        tea_temperature -> Nullable<Integer>,
        tea_kind -> Nullable<Text>,
        coffee_roast -> Nullable<Text>,
        coffee_origin -> Nullable<Text>,
    }
}
```

For a setup like this, it is not possible to implement `ToSql` and `FromSql` for `TeaOrCoffee`, as its constituent values should end up in two separate columns. What would be the way to achieve this?

**Comment by weiznich** [comment, marked as answer]
In this kind of setup its not possible to work only with the derives. You need to manually implement `Insertable` and `Queryable` for some types. The good thing is that this isn't particular hard, the code is just a bit repetitive, as those two traits just map struct fields to tuples. 

For `Insertable` it's necessary to provide two implementations of that trait for the `TeaOrCoffee` enum, one for the owned type and one for references to the owned type. In addition it's necessary to derive `Insertable` for the `Tea` and `Coffee` structs as well. The relevant impl for `TeaOrCoffee` looks like that:

```rust
impl Insertable<menus::table> for TeaOrCoffee {
    type Values = <(Option<Tea>, Option<Coffee>) as Insertable<menus::table>>::Values;

    fn values(self) -> Self::Values {
        match self {
            TeaOrCoffee::Tea(tea) => (Some(tea), None::<Coffee>).values(),
            TeaOrCoffee::Coffee(coffee) => (None::<Tea>, Some(coffee)).values(),
        }
    }
}
```

It essentially maps your enum to a tuple of options of the inner types. (There is a second impl for `&TeaOrCoffee` needed, that just uses references everywhere). This works because `#[derive(Insertable)]` supports `#[diesel(embed)]`

For `Queryable` there are several solutions. It mostly depends on what your query is supposed to return. I decided to go with the simplest possible case, that one that your query returns all the columns of your table as unested column. In this case we need to implement `Queryable` manually for the top level `Menu` type to map the returned value tuple to the corresponding type structure:

```rust
impl<ST> Queryable<ST, Sqlite> for Menu
where
    (
        i32,
        String,
        Option<i32>,
        Option<String>,
        Option<String>,
        Option<String>,
    ): FromStaticSqlRow<ST, Sqlite>,
{
    type Row = (
        i32,
        String,
        Option<i32>,
        Option<String>,
        Option<String>,
        Option<String>,
    );

    fn build(
        (_id, food, tea_temperature, tea_kind, coffee_roast, coffee_origin): Self::Row,
    ) -> diesel::deserialize::Result<Self> {
        let beverage = match (tea_temperature, tea_kind, coffee_roast, coffee_origin) {
            (Some(temperature), Some(kind), None, None) => {
                TeaOrCoffee::Tea(Tea { temperature, kind })
            }
            (None, None, Some(roast), Some(origin)) => {
                TeaOrCoffee::Coffee(Coffee { roast, origin })
            }
            _ => return Err("Invalid database state".into()),
        };
        Ok(Self { food, beverage })
    }
}
```

(This implementation now targets `Sqlite` but you can simply use any other database type there by replacing the relevant type.)

I've tested the whole implementation with the code in [this gist](https://gist.github.com/weiznich/0d808bbb461e75ed232167200ad16438)

**Reply by dspicher** [reply]
That's great, thank you so much for taking the time!
